src/tests/suite/dfs_unit_test.c

```diff
@@ -0,0 +1,294 @@
+/**
+ * (C) Copyright 2019-2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+#include "daos_iotest.h"
+#include <pthread.h>
+#include <daos_fs.h>
+
+/** global DFS mount used for all tests */
+static uuid_t           co_uuid;
+static daos_handle_t    co_hdl;
+static dfs_t		*dfs_mt;
+
+static void
+dfs_test_mount_umount(void **state)
+{
+	test_arg_t		*arg = *state;
+	uuid_t			cuuid;
+	daos_cont_info_t	co_info;
+	daos_handle_t		coh;
+	dfs_t			*dfs;
+	int			rc;
+
+	if (arg->myrank != 0)
+		return;
+
+	/** create & open a non-posix container */
+	uuid_generate(cuuid);
+	rc = daos_cont_create(arg->pool.poh, cuuid, NULL, NULL);
+	assert_int_equal(rc, 0);
+	print_message("Created non-POSIX Container "DF_UUIDF"\n",
+		      DP_UUID(cuuid));
+	rc = daos_cont_open(arg->pool.poh, cuuid, DAOS_COO_RW,
+			    &coh, &co_info, NULL);
+	assert_int_equal(rc, 0);
+
+	/** try to mount DFS on it, should fail. */
+	rc = dfs_mount(arg->pool.poh, coh, O_RDWR, &dfs);
+	assert_int_equal(rc, EINVAL);
+
+	rc = daos_cont_close(coh, NULL);
+	assert_int_equal(rc, 0);
+	rc = daos_cont_destroy(arg->pool.poh, cuuid, 1, NULL);
+	assert_int_equal(rc, 0);
+	print_message("Destroyed non-POSIX Container "DF_UUIDF"\n",
+		      DP_UUID(cuuid));
+
+	/** create a DFS container with POSIX layout */
+	rc = dfs_cont_create(arg->pool.poh, cuuid, NULL, NULL, NULL);
+	assert_int_equal(rc, 0);
+	print_message("Created POSIX Container "DF_UUIDF"\n", DP_UUID(cuuid));
+	rc = daos_cont_open(arg->pool.poh, cuuid, DAOS_COO_RW,
+			    &coh, &co_info, NULL);
+	assert_int_equal(rc, 0);
+
+	rc = dfs_mount(arg->pool.poh, coh, O_RDWR, &dfs);
+	assert_int_equal(rc, 0);
+
+	rc = dfs_umount(dfs);
+	assert_int_equal(rc, 0);
+	rc = daos_cont_close(coh, NULL);
+	assert_int_equal(rc, 0);
+	rc = daos_cont_destroy(arg->pool.poh, cuuid, 1, NULL);
+	assert_int_equal(rc, 0);
+	print_message("Destroyed POSIX Container "DF_UUIDF"\n", DP_UUID(cuuid));
+
+	/** create a DFS container with POSIX layout */
+	rc = dfs_cont_create(arg->pool.poh, cuuid, NULL, NULL, NULL);
```

Comment:
there is no need to create, open, and destroy a container on every test.
you can reuse the global container uuid that is created in setup (co_hdl) and avoid all container operations to improve the code.